### ISSUE TITLE
Support for Treadmill devices

### ISSUE THREAD

**Issue by AgentMilkshake1** [ValveSoftware/openvr#937]
Hi there,

I've written about this before, but I'm still to receive any response from the development team on a timeline for treadmill device support.

Currently it seems that the new input binding system only allows for 2 controllers (left and right motion controller) for any 1 HMD.

Are there plans in the near future to add support for a third input device. This would allow a treadmill device trackpad/thumbstick input to project onto a left or right motion controllers trackpad/thumbstick whilst still being able to use both motion controllers at the same time.

This problem has been identified by many developers and we are still yet to see any feedback on the matter.

#923 
#886
#856
#826

This is clearly possible as we are able to send a gamepad's input to motion controllers trackpad/thumbstick. Could we please get a patch soon to allow this as it is preventing many VR devices from working with Steam VR.

Many thanks

**Comment by kbigdelysh** [comment]
You can write a driver for OpenVR and achieve that (I've written a driver for our own custom controllers). It's  not that hard. There is a sample driver [here] (https://github.com/ValveSoftware/openvr/tree/master/samples/driver_sample).

**Comment by AgentMilkshake1** [comment]
Hi @kbigdelysh ,

I've written a driver for a custom device using this exact sample, but games/apps only bind to this input when one of the motion controllers are turned off (the one you are binding to) or if you give your controller a higher priority (in which case you cannot use any of the inputs on the lower priority motion controller).

 There are many developers who have tried doing the same thing so if you could provide a more thorough example that would be great.

Thanks

**Comment by kbigdelysh** [comment]
We had similar issue. Make sure your controller is not occupying right or left controller role. In the null driver example, make sure for you custom controller change the following: 

> vr::VRProperties()->SetInt32Property(m_ulPropertyContainer, Prop_ControllerRoleHint_Int32, **TrackedControllerRole_RightHand**)

to:

> vr::VRProperties()->SetInt32Property(m_ulPropertyContainer, Prop_ControllerRoleHint_Int32, **TrackedControllerRole_OptOut**)

This change solved our issue.

**Comment by AgentMilkshake1** [comment]
Thanks for this feedback. Unfortunately this is exactly how the project has been setup for months now, and it doesn't work.

Was it only that specific change for you that got it working?

**Comment by kbigdelysh** [comment]
For me, yes but also make sure your json files in 
C:\Program Files (x86)\Steam\steamapps\common\[Your-Driver]\resources\input\
have been modified correctly.

**Comment by AgentMilkshake1** [comment]
They are, Steam VR receives all of the input and I can confirm that if I were to change it to left or right hand, and then turn off said motion controller, Steam VR binds the custom device to the game.

Unfortunately, when the device is marked as opt out, SteamVR detects the input (and it shows up in the input debugger) but games do not bind.

This isn't my first rodeo :P

**Comment by JoeLudwig** [comment]
The latest SDK adds the start of treadmill support: https://github.com/ValveSoftware/openvr/releases/tag/v1.1.3

Once your treadmill starts coming in as /user/treadmill it should work well in the binding UI. The game will receive analog values for whatever x/y values are furthest from 0,0, so I would expect the treadmill and trackpad to work at the same time.

**Comment by AgentMilkshake1** [comment]
Hey @JoeLudwig , 

Thanks for the update to the SDK, I'll be checking out the changes soon. Ironically it comes days after finishing implementation of our driver as a generic tracker with scalar input components! haha

I'm guessing it would probably be ideal for us to transition over to this new implementation?

**Comment by JoeLudwig** [comment]
It will probably help, but give it a try and tell me what you think. 

The treadmill role currently requires that the device in question be a controller. I'd rather stick with that unless it turns out to cause a bunch of problems.

**Comment by AgentMilkshake1** [comment]
Sounds good to me - I've already made the changes back to controller and added the specifier as treadmill (controller role hint right?) - just waiting on the compiler issue for openvr_driver.h which I know you've already seen and assigned.

I'll let you know how my testing goes once that change has been submitted to openvr_driver.h

**Comment by JoeLudwig** [comment]
We'll try to get the driver file fixed soon. In the meantime you can hack a fix into the file yourself.

**Comment by AgentMilkshake1** [comment]
From my limited tests so far, it works exactly as expected. Thank you so much for working on this feature.

2 things:
1) Can I suggest for the near future that we leave this issue open, just in case any new bugs crop up, and to give much more visibility for developers looking for this.

2) Would you like me to write a short "How to document" for creating a treadmill like device driver? I could write it on Google Docs in the next couple of days, link it here for you to look over, and then you could copy paste it onto the OpenVR wiki for community reference?

**Comment by Skaarah** [comment]
Please, I would love to see a working implementation. I still haven't got my device to work.

**Comment by Skaarah** [comment]
I managed to get it working! I missed that there was a new controller role for treadmill. Now my software finally works.

**Comment by AgentMilkshake1** [comment]
@Skaarah Fantastic, great to hear!

**Comment by JoeLudwig** [comment]
@Skaarah @AgentMilkshake1 @kbigdelysh @SamRelativeDimensions

Where do we stand with overall treadmill support now that we have the treadmill controller role? What remains to make treadmills be a smooth experience for users?

**Comment by cwlmyjm** [comment]
I have done a sample with new openvr, trackpad is workable. But fail on creating system click component, wiki said that component is not available to applications. I have tried to create binding file  to bind that component to /actions/legacy/in/left_system_press (also try right_system_press) like what vive controller does but not work. Inside input debugger, it get the signal but dashboard ot summon.

**Comment by AgentMilkshake1** [comment]
@JoeLudwig I've been speaking with some developers who use/intend to use the treadmill functionality. There has been some interest in whether it would be possible to allow game developers to utilise all 3 "thumbstick" paths for true treadmill support.

For instance, game developers might want to setup a treadmill mode in their game where the treadmill thumbstick input feeds into the movement of the character, the right controller thumbstick/trackpad feeds into rotating the character, and the left thumbstick/trackpad acts as a radial menu.

I'm not sure how possible this is (would probably require some new features in both UE4 and Unity's SteamVR SDK) but there is some interest.

The only other mention I've seen so far about things not working as expected is cwlmyjm's post above.

**Comment by peroht** [comment]
@JoeLudwig thanks guys for this implementation - we've (Omnifinity) have been waiting for this for a long time before the inception of the Vive and the initial efforts on this topic in OSVR. Glad this work is taken further in the Khronos Group and we do not have to hack/overwrite device values using the fantastic work done by matzman666 in [https://github.com/matzman666/OpenVR-InputEmulator](https://github.com/matzman666/OpenVR-InputEmulator).

> What remains to make treadmills a smooth experience for users?

This is how I see a path forward - please feel free to comment on how you see it @Skaarah @AgentMilkshake1 @kbigdelysh @SamRelativeDimensions. 

The first step toward this is now implemented in openvr/steamvr - this is greate news. 
As a second step this means that we - as hardware manufacturers - can begin to adress our fragmentation of how character movement in games is implemented. 
The thirds step is for us to agree on a set of typical scenarios/recommendations on how game developers should go about when implementing locomotion driven by our various locomotion devices. 
The forth step is not rocket science. It basically constitutes of us manufacturers being a unified force. This will hopefully make it easier and more interesting for game developers to implement treadmill locomotion in their standard locomotion options. This is something that I belive has stifled support due to this being super niche. 

From the perspective of the Omnideck a game developer using e.g. Unity would typically implement the simplest form of character movement by moving the [CameraRig] using X/Y Axis analogue values representing [m/s]. It is clear to see that the key for us is to match the "real" locomotion speed of the human with the virtual speed in the game. For this reason we do not recommend e.g. discrete input such as WASD on our devicer or excessive movement multiplication factors.

Today, in games like Onward which allows the player to alter/rotate their forward direction. For the Omnideck this is a negative and undesired functionality. The same goes for snap turning. As I see it, we should - for each locomotion device - recommend to the game developers on how to best implement support for each device, just as @AgentMilkshake1 describes in the post above. 
As a reference, the game "Talos Principle" have got a locomotion called "Fixed Playspace" that represents how our device would control the game. The calibrated playspace and the in-game coordinate systems are always aligned with no hand controllers affecting the camera/character rotation.

Thanks.

**Comment by peroht** [comment]
@JoeLudwig I've managed to dive into this on a technical level. 

With the driver I am getting a "/user/treadmill -omnideck_controller" field in the input debugger. I see that my various input component paths are receiving data. What I do not understand is which "input component path" I should define for the scalar/boolean x and y components? Ive tried a normal "/input/trackpad/x" and also a "user/treadmill/x" and see that data is arriving, but has no effect.

I do not understand how I should configure the above path/data that my "/user/treadmill" is controlling to replace e.g. the trackpad of a left controller which uses smooth locomotion in various games. I am using legacy games at the moment and have read the big documentation [https://steamcommunity.com/sharedfiles/filedetails/?id=1545634111](https://steamcommunity.com/sharedfiles/filedetails/?id=1545634111) but do not really see how to tread forward to make games accept data from the custom device.

Have a great day.

**Comment by AgentMilkshake1** [comment]
Hey everyone!

A new potential feature request for Treadmill devices - allowing for input "curve editing" of the joystick component in the same way that you can edit the deadzone etc. for Gamepad's in Big Picture Mode.

@JoeLudwig Is this something we could expect to see in the near future? Thanks!

**Comment by AgentMilkshake1** [comment]
Hi all,

Myself and other developers have recently started noticing that Treadmill support has been compromised in recent updates.

I've personally not been very active with OpenVR development over the past few months, but I've recently had reports that since the Valve Index has launched many games have updated their SteamVR input systems (a requirement to support Valve Index).
As a part of this, game developers are now able to filter actions depending on source (right hand, left hand, any). This means that actions created by game developers are only usable by Treadmill devices in the case that the game developer chooses ANY.
This is a massive step back for Treadmill developers, as many game developers won't realise the impact their choice at dev level is having at the wider market level. A lot of games previously supported by treadmills are beginning to dwindle in numbers due to this oversight.

We had a brief period where Treadmill devices were becoming quite universal in terms of compatibility, but these recent updates have really damaged the support Treadmill device developers can offer!

@JoeLudwig What is being done by the OpenVR team to remedy this? Were the team aware that this would be a problem?

**Comment by JoeLudwig** [comment]
This issue is giant and mostly has nothing to do with SteamVR Input games filtering input by device. Please open a new issue.
